File: MovieQuiz/MovieGrabber.m
//
//  MovieGrabber.m
//  MovieQuiz
//
//  Created by Renan Greca on 3/18/15.
//  Copyright (c) 2015 Renan Greca. All rights reserved.
//

#import "MovieGrabber.h"
#import "Movie.h"

@interface MovieGrabber ()
@end

@implementation MovieGrabber

+ (BOOL) put:(int)limit moviesInList:(NSMutableArray *)list {
    //MovieList *list = [MovieList sharedInstance];

    NSString *apiKey = @"asprbevazhnusm6fjwqnk24d";
    
    NSURL *url = [NSURL URLWithString:[NSString stringWithFormat:@"http://api.rottentomatoes.com/api/public/v1.0/lists/dvds/new_releases.json?apikey=%@&page_limit=%d", apiKey, limit]];
    //NSString *apiUrl = [NSString stringWithFormat:@"http://api.rottentomatoes.com/api/public/v1.0/lists/dvds/new_releases.json?apikey=%@&page_limit=%d", apiKey, limit];
    
    NSError *error = nil;
    NSData *urlData = [NSData dataWithContentsOfURL:url];
    
    if (!urlData) {
        return NO;
    }
    
    NSDictionary *data = [NSJSONSerialization JSONObjectWithData:urlData options:0 error:&error];
    
    if ([data objectForKey:@"error"]) {
        return NO;
    }
    
    NSArray *movies = [data objectForKey:@"movies"];
    
    for (NSDictionary *movie in movies) {
        NSString *title = [movie objectForKey:@"title"];
        
        NSString *synopsis = [movie objectForKey:@"synopsis"];
        synopsis = [MovieGrabber filter:synopsis with:title];
        
        NSString *img = [[movie objectForKey:@"posters"] objectForKey:@"detailed"];
        
        [list addObject:[Movie MovieWithTitle:title andSynopsis:synopsis andImg:img]];
    }
        
    //Movie *movie = [list objectAtIndex:0];
    return YES;
}

+ (NSString *) getDataFrom:(NSString *)url{
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
    [request setHTTPMethod:@"GET"];
    [request setURL:[NSURL URLWithString:url]];
    
    NSError *error = [[NSError alloc] init];
    NSHTTPURLResponse *responseCode = nil;
    
    NSData *oResponseData = [NSURLConnection sendSynchronousRequest:request returningResponse:&responseCode error:&error];
    
    if([responseCode statusCode] != 200){
        NSLog(@"Error getting %@, HTTP status code %li", url, (long)[responseCode statusCode]);
        return nil;
    }
    
    return [[NSString alloc] initWithData:oResponseData encoding:NSUTF8StringEncoding];
}

+ (NSString *) getiTunesURLWithTitle:(NSString *)title {
    title = [title stringByReplacingOccurrencesOfString:(NSString *)@" "
                                            withString:(NSString *)@"+"
                                            options:NSCaseInsensitiveSearch
                                            range:NSMakeRange(0, [title length])];
    
    NSURL *url = [NSURL URLWithString:[NSString stringWithFormat:@"https://itunes.apple.com/search?entity=movie&term=%@", title]];

    NSError *error = nil;
    NSData *urlData = [NSData dataWithContentsOfURL:url];

    NSDictionary *data = [NSJSONSerialization JSONObjectWithData:urlData options:0 error:&error];

    NSString *iTunesURL = [[[data objectForKey:@"results"] objectAtIndex:0] objectForKey:@"trackViewUrl"];
    
    return iTunesURL;
}

+ (NSString *) filter:(NSString *)text with:(NSString *)title {

    text = [text stringByReplacingOccurrencesOfString:(NSString *)title
                                            withString:(NSString *)@"_____"
                                            options:NSCaseInsensitiveSearch
                                            range:NSMakeRange(0, [text length])];
    
    return text;
}

@end

File: MovieQuiz/MovieGrabber.swift
//
//  MovieGrabber.swift
//  MovieQuiz
//
//  Created by Renan Greca on 5/6/15.
//  Copyright (c) 2015 renangreca. All rights reserved.
//

import Foundation
import SwiftyJSON

func fetchMovies(limit: Int) -> Array<Movie> {    
    // create the request
    let url = URL(string: "https://imdb-api.com/en/API/MostPopularMovies/\(APIKeys.IMDb)")
    
    var returnMovies: Array<Movie> = []

    if let data = try? Data(contentsOf: url!),
       let json = try? JSON(data: data),
       let movies = json["items"].array {
        var i = 0
        while i < limit {
            if let movie = Movie(json: movies[i]) {
                returnMovies.append(movie)
                i += 1
            }
        }
    } else {
        print("Error fetching API data.")
    }
    
    return returnMovies
}

func getURLsAndImages(movies: Array<Movie>) {
    for movie in movies {
        getStoreURL(movie: movie)
        getOMDbImg(movie: movie)
        getImageFromURL(movie: movie)
    }
}

func getURLsAndImages(movie: Movie) {
    getStoreURL(movie: movie)
    getOMDbImg(movie: movie)
    getImageFromURL(movie: movie)
}

func getStoreURL(movie: Movie) {
    if movie.itunesURL == "" {
        let cleanTitle = movie.title.replaceAll(find: " ", with: "+")
        let url = URL(string: "https://itunes.apple.com/search?entity=movie&term=\(cleanTitle)")
        let data = try? Data(contentsOf: url!)
        
        if data == nil {
            print("iTunes 💩")
            return;
        }
        
        let json = try? JSON(data: data!)
        
        movie.itunesURL = json!["results"][0]["trackViewUrl"].stringValue
        movie.itimgURL = json!["results"][0]["artworkUrl100"].stringValue
    }
}

func getOMDbImg(movie: Movie) {
    if (movie.img == UIImage(named: "MovieQuizLogo")) {
        let url = URL(string: "http://www.omdbapi.com/?i=\(movie.imdbID)&plot=short&r=json&apikey=\(APIKeys.OMDb)")
        let data = try? Data(contentsOf: url!)
        
        if data == nil {
            print("OMDb 💩")
            return
        }
        
        let json = try? JSON(data: data!)
        
        movie.imgURL = json!["Poster"].stringValue
        
        let imgdata = try? Data(contentsOf: URL(string: movie.imgURL)!)
        if let imgdata = imgdata {
            if let img = UIImage(data: imgdata) {
                movie.img = img
            }
        }
        
        movie.synopsis = json!["Plot"].stringValue
    }
}

func getImageFromURL(movie: Movie) {
    if (movie.img == UIImage(named: "MovieQuizLogo")) {
        let url = URL(string: movie.imgURL)
        let data = try? Data(contentsOf: url!)
        
        if data == nil {
            print("RT img 💩")
            return
        }
        
        if let img = UIImage(data: data!) {
            movie.img = img
        }
    }
}





